numpy/lib/function_base.py

```diff
@@ -2996,9 +2992,33 @@ def _median(a, axis=None, out=None, overwrite_input=False):
         indexer[axis] = slice(index, index+1)
     else:
         indexer[axis] = slice(index-1, index+1)
-    # Use mean in odd and even case to coerce data type
-    # and check, use out array.
-    return mean(part[indexer], axis=axis, out=out)
+
+    # Check if the array contains any nan's
+    if np.issubdtype(a.dtype, np.inexact):
+        # warn and return nans like mean would
+        rout = mean(part[indexer], axis=axis, out=out)
+        part = np.rollaxis(part, axis, part.ndim)
+        n = np.isnan(part[..., -1])
+        if rout.ndim == 0:
+            if n == True:
+                warnings.warn("Invalid value encountered in median",
+                              RuntimeWarning)
+                if out is not None:
+                    out[...] = a.dtype.type(np.nan)
+                    rout = out
+                else:
+                    rout = a.dtype.type(np.nan)
+        else:
+            for i in range(np.count_nonzero(n.ravel())):
+                warnings.warn("Invalid value encountered in median",
```

Comment:
Issuing warnings in a loop seems like a bad idea, for two reasons:

1. It would just produce a lot of noise in logs (if anyone is not filtering all but the first warning)
2. Issuing warnings is actually very slow, so this could be a serious performance bottleneck. Consider:

```
In [10]: a = np.zeros((10000, 2))

In [11]: %timeit np.median(a, axis=0)
10000 loops, best of 3: 117 µs per loop

In [12]: %timeit for x in range(10000): warnings.warn("Invalid value encountered in median", RuntimeWarning)
/Users/shoyer/miniconda/envs/numpy-dev/bin/ipython:257: RuntimeWarning: Invalid value encountered in median
100 loops, best of 3: 12.3 ms per loop
```